docs/03-channels.html

```diff
@@ -0,0 +1,335 @@
+  end
+end
+-- spawn the task to print tick every two seconds
+cosock.spawn(function()
+  task(tock_tx, tick_rx, &quot;tick&quot;)
+end)
+-- spawn the task to print tock every 2 seconds
+cosock.spawn(function()
+  task(tick_tx, tock_rx, &quot;tock&quot;)
+end)
+-- prime the tick task tp start first
+tick_tx:send()
+
+cosock.run()
+</code></pre>
+<p>In this version, we only have one definition for a task, looping forever, it will first wait to <code>receive</code> the signal,
+and then it prints its name, sleeps for 1 second and then sends the signal on to the next task. The key to 
+making this all work is that we need to kick the process off by telling one of the tasks to start. Since a
+channel allows for multiple senders, it is ok that we call <code>send</code> in more than one place.</p>
+<p>Now say we wanted to extend our little clock emulator to print <code>&quot;clack&quot;</code> every 60 seconds to simulate the
+minute hand moving. That might look something like this:</p>
+<pre><code class="language-lua">{{#include ../example/tick_tock_clack.lua}}
+</code></pre>
+<p>Here we have updated the tasks to now share a counter across our channel. So at the 
+start of each loop iteration, we first get the current count from our other task. We again print our name
+and then sleep for 1 second but now if the count is &gt;= 59 we send the count of 0 to our <code>&quot;clack&quot;</code> task which
+will always then send a 1 to the <code>&quot;tick&quot;</code> task to start the whole process over again. Just to make sure it
+is clear, we can use the send half of the <code>&quot;tick&quot;</code> task's channel in 3 places, the main thread
+to &quot;prime&quot; the clock, the <code>&quot;tock&quot;</code> task and the <code>&quot;clack&quot;</code> task.</p>
+<p>It is very important that we don't try and use the receiving half of a channel in more than one
+task, that would lead to potentially unexpected behavior. Let's look at an example of how that might
+go wrong.</p>
+<pre><code class="language-lua">-- bad_news_channels.lua
+local cosock = require &quot;cosock&quot;
+
+local tx, rx = cosock.channel.new()
+
+cosock.spawn(function()
+  rx:receive()
+  print(&quot;task 1&quot;)
+end)
+
+cosock.spawn(function()
+  rx:receive()
+  print(&quot;task 2&quot;)
+end)
+
+tx:send()
+
+cosock.run()
+</code></pre>
+<p>In this example, we create one channel pair and spawn two tasks which both call <code>receive</code> on our
+channel and just before <code>run</code> we call <code>send</code>.  Since the choice for which task
+should run at which time is left entirely up to cosock, we can't say for sure which of these tasks
+will actually receive. It might print &quot;task 1&quot; then &quot;task 2&quot; or it might print them in the reverse. </p>
+<blockquote>
+<p>In actuality, cosock assumes that <code>receive</code> will only ever be called from the same coroutine.
+Calling <code>receive</code> in multiple coroutintes will (eventually) raise in a error.</p>
```

Comment:
This is not book related, but cosock related.  The error message here is:
```
lua: /usr/local/share/lua/5.3/cosock.lua:294: cosock tried to call socket.select with no sockets and no timeout. this is a bug, please report it
stack traceback:
        [C]: in function 'error'
        /usr/local/share/lua/5.3/cosock.lua:294: in function 'cosock.run'
        ./bad_news_channels.lua:18: in main chunk
        [C]: in ?
```

Where it seems like this is UE not a bug that should be reported.

Reply:
I spent a good amount of time this weekend trying to unwrap this to allow for `mpsc` semantics in the `BoundedChannel` example. I actually think this might be an actual bug since I would expect each thread would get it's own entry in `theadswaitingfor` and somehow this isn't actually happening. I think there is some special handling in the `cosock.socket` tables for raising an error if a `receiver` is used in multiple threads but I need to dig a little to find that